### PR TITLE
feat: add acurast hyperdrive pallet

### PR DESCRIPTION
**Pull request by RomarQ**
## Summary

create a `pallet_acurast_hyperdrive` to support external chain state sync and add support for the Tezos chain proof validation.

## What is the expected behavior?

- submit merkle root extrisic (snapshot number (uint), merkle root (bytes))
- submit proof extrinsic (key/value, proof path, snapshot number, nonce)
  - requires a nonce

### CODE REVIEW

File: pallets/hyperdrive/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+            block: T::TargetChainBlockNumber,
+            state_merkle_root: T::TargetChainHash,
+        },
+        StateMerkleRootAccepted {
+            block: T::TargetChainBlockNumber,
+            state_merkle_root: T::TargetChainHash,
+        },
+    }
+
+    /// This storage field maps the state transmitters to their respective activiti window.
```
Comment (RomarQ) [inline pallets/hyperdrive/src/lib.rs:106]: My typo
```suggestion
    /// This storage field maps the state transmitters to their respective activity window.
```

File: pallets/hyperdrive/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+                block
+                    .checked_rem(&T::TransmissionRate::get())
+                    .ok_or(Error::<T, I>::CalculationOverflow)?
+                    .is_zero(),
+                Error::<T, I>::SubmitOutsideTransmissionRate
+            );
+
+            // insert merkle root proposal since all checks passed
+            // allows for constant-time validity checks
+            let accepted =
```
Comment (RomarQ) [inline pallets/hyperdrive/src/lib.rs:218]: The idea here was to expect at least X root submissions and then get the one with the majority of the votes. But this works for now.

File: pallets/hyperdrive/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+                });
+            }
+
+            Ok(())
+        }
+    }
+
+    pub enum ValidationResult {
+        UnconfirmedStateRoot,
+    }
```
Comment (RomarQ) [inline pallets/hyperdrive/src/lib.rs:244]: Is this being used?

File: pallets/hyperdrive/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+            + Debug
+            + MaybeDisplay
+            + SimpleBitOps
+            + Ord
+            + Default
+            + Copy
+            + CheckEqual
+            + sp_std::hash::Hash
+            + AsRef<[u8]>
+            + AsMut<[u8]>
```
Comment (RomarQ) [inline pallets/hyperdrive/src/lib.rs:56]: I do not think we need all these constraints.

File: pallets/hyperdrive/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+            + MaybeDisplay
+            + AtLeast32BitUnsigned
+            + Default
+            + Bounded
+            + Copy
+            + sp_std::hash::Hash
+            + FromStr
+            + MaxEncodedLen
+            + TypeInfo
+            + Zero
```
Comment (RomarQ) [inline pallets/hyperdrive/src/lib.rs:72]: I do not think we need all these constraints.

File: pallets/hyperdrive/src/lib.rs (hunk 4 above)
Comment (gitsimon) [reply pallets/hyperdrive/src/lib.rs:56]: I took them from frame_system::Config and only added extra what I need. We will see what we need after implementing Hash derivation.

File: pallets/hyperdrive/src/lib.rs (hunk 5 above)
Comment (gitsimon) [reply pallets/hyperdrive/src/lib.rs:72]: I took them from frame_system::Config and only added extra what I need. We will see what we need.

File: pallets/hyperdrive/src/lib.rs (hunk 2 above)
Comment (gitsimon) [reply pallets/hyperdrive/src/lib.rs:218]: the current implementation assumes we make sure quorum size is > number of ceil(transmitters/2)

**Review by gitsimon — APPROVED** [review]

**Pull request merged**